tools/fst/prepare_dict.py

```diff
@@ -6,41 +6,56 @@
 # sys.argv[1]: e2e model unit file(lang_char.txt)
 # sys.argv[2]: raw lexicon file
 # sys.argv[3]: output lexicon file
+# sys.argv[4]: bpemodel
 
 unit_table = set()
 with open(sys.argv[1], 'r', encoding='utf8') as fin:
     for line in fin:
         unit = line.split()[0]
         unit_table.add(unit)
 
+
+def contain_oov(units):
+    for unit in units:
+        if unit not in unit_table:
+            return True
+    return False
+
+
 lexicon_table = set()
 with open(sys.argv[2], 'r', encoding='utf8') as fin, \
-     open(sys.argv[3], 'w', encoding='utf8') as fout:
+        open(sys.argv[3], 'w', encoding='utf8') as fout:
+    bpemode = len(sys.argv) > 4
+    if bpemode:
+        import sentencepiece as spm
+        sp = spm.SentencePieceProcessor()
+        sp.Load(sys.argv[4])
     for line in fin:
         word = line.split()[0]
-        if word == 'SIL':
+        if word == 'SIL' and not bpemode:  # `sil` might be a valid piece in bpemodel
             continue
         elif word == '<SPOKEN_NOISE>':
             continue
         else:
             # each word only has one pronunciation for e2e system
             if word in lexicon_table:
                 continue
-            # ignore words with OOV
-            contains_oov = False
-            for ch in word:
-                if ch not in unit_table:
-                    contains_oov = True
-                    break
-            if contains_oov:
-                print(
-                    'Ignoring words {}, which contains oov unit'.format(word))
-                continue
-            # Optional, append ▁ in front of english word
-            # we assume the model unit of our e2e system is char now.
-            # TODO(Binbin Zhang): support BPE by sentencepiece
-            if word.encode('utf8').isalpha() and '▁' in unit_table:
-                word = '▁' + word
-            chars = ' '.join(word)  # word is a char list
+            if bpemode:
+                pieces = sp.EncodeAsPieces(word)
+                if contain_oov(pieces):
+                    print('Words {} contains oov unit, replace oov unit with \"<unk>\"'.format(
+                        ''.join(word).strip('▁')))
+                chars = ' '.join(
+                    [p if p in unit_table else '<unk>' for p in pieces])
```

Comment:
words with oov unit must be included in lexicon as long as we use downloaded lm, otherwise, TLG.fst cannot be built (it throws errors like `FstCompiler: Symbol "ZZZ" is not mapped to any integer arc ilabel`, in which "ZZZ" contains oov piece).

Reply:
okay, I see.